pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,487 @@
+		return existingTv
+	}
+	return tv
+}
+
+// storeEntry stores the TracedEntry without a tracing context and returns the
+// newly created traceValue.
+func (r *RaftTracer) storeEntry(te kvserverpb.TracedEntry) *traceValue {
+	tv := traceValue{traced: te}
+	tv.ctx = logtags.AddTag(r.ctx, "id", redact.Safe(tv.String()))
+	tv.mu.seenMsgAppResp = make(map[raftpb.PeerID]bool)
+	return r.tryStore(&tv)
+}
+
+// storeEntryWithTracing stores the TracedEntry along with a tracing context and
+// returns the newly created traceValue.
+func (r *RaftTracer) storeEntryWithTracing(
+	propCtx context.Context, span *tracing.Span, te kvserverpb.TracedEntry,
+) *traceValue {
+	tv := traceValue{traced: te}
+	tv.ctx = logtags.AddTag(r.ctx, "id", redact.Safe(tv.String()))
+	tv.mu.propCtx = propCtx
+	tv.mu.propSpan = span
+	tv.mu.seenMsgAppResp = make(map[raftpb.PeerID]bool)
+	return r.tryStore(&tv)
+}
+
+// RegisterRemote registers a remote trace. This is called when we receive a
+// raft message over the wire with a request to continue tracing it.
+func (r *RaftTracer) RegisterRemote(te kvserverpb.TracedEntry) {
+	if !r.reserveSpace() {
+		return
+	}
+	// NB: We don't currently return remote traces, if we did, we would pass the
+	// remote ctx here and trace it. The problem is knowing when to send it
+	// back to the remote node.
+	tv := r.storeEntry(te)
+	tv.logf(1, "registering remote trace %s", tv)
+}
+
+// MaybeRegister is called on an entry that has been proposed to raft. This will
+// begin logging all subsequent updates to this entry. It returns true if the
+// registration is successful. A duplicate registration of the same index is
+// considered a success and returns true, however the older registration is kept
+// and this registration is ignored.
+func (r *RaftTracer) MaybeRegister(ctx context.Context, ent raftpb.Entry) bool {
+	// If the index is nil, then we can't trace this entry. This can happen if
+	// there is a leader/leaseholder spilt. We don't have an easy way to handle
+	// this today, so don't attempt to trace it.
+	if ent.Index == 0 {
+		log.VEvent(ctx, 2, "skip registering raft proposal without index")
+		return false
+	}
+
+	// Only register the entry if this is a traced context with verbose logging.
+	span := tracing.SpanFromContext(ctx)
+	if span == nil || span.RecordingType() != tracingpb.RecordingVerbose {
+		return false
+	}
+
+	// This must be the last conditional. If this returns true we must call
+	// storeEntryWithTracing to not leak a registered permit.
+	if !r.reserveSpace() {
+		log.VEvent(ctx, 2, "too many active raft traces, skipping")
+		return false
+	}
+
+	ctx, span = r.tracer.StartSpanCtx(ctx, "raft trace",
+		tracing.WithParent(span), tracing.WithFollowsFrom())
+
+	tv := r.storeEntryWithTracing(
+		ctx,
+		span,
+		kvserverpb.TracedEntry{
+			Index:   kvpb.RaftIndex(ent.Index),
+			TraceID: span.TraceID(),
+			SpanID:  span.SpanID(),
+		},
+	)
+	tv.logf(1, "registering local trace %s", tv)
```

Comment:
`storeEntryWithTracing` can return an old trace, right? We probably don't want to log this "registering" message again in this case. Consider `storeEntryWithTracing` to return the indicator whether `tv` is new, and only log in that case. If it's not new, return false.

Same applies to the `RegisterRemote` call.

Reply:
Done.